### ISSUE TITLE
Typesense hang out after few days (can't reset_peer)

### ISSUE THREAD

**Issue by jasiek-net** [typesense/typesense#904]
## Description

After few days typesense response for every request with `{ "message": "Not Ready or Lagging" }`

## Steps to reproduce

Just run typesense for few days.

## Expected Behavior

Not lagging.

## Actual Behavior

Lagging, logs from docker below:

```
I20230222 16:46:05.087571    84 raft_server.cpp:534] Term: 2, last_index index: 242, committed_index: 241, known_applied_index: 241, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 721
W20230222 16:46:05.087608    84 raft_server.cpp:558] Single-node with no leader. Resetting peers.
W20230222 16:46:05.087615    84 node.cpp:866] node default_group:172.17.0.2:8107:8108 is in state ERROR, can't reset_peer
```

## Metadata

**Typesense Version**: typesense/typesense:0.23.1

**OS**: 
```
Architektura:            x86_64
Tryb(y) pracy CPU:       32-bit, 64-bit
Kolejność bajtów:        Little Endian
CPU:                     1
Lista aktywnych CPU:     0
Wątków na rdzeń:         1
Rdzeni na gniazdo:       1
Gniazd:                  1
Węzłów NUMA:             1
ID producenta:           GenuineIntel
Rodzina CPU:             6
Model:                   79
Nazwa modelu:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Wersja:                  1
CPU MHz:                 2299.925
BogoMIPS:                4600.14
Producent hipernadzorcy: Xen
Typ wirtualizacji:       pełna
Cache L1d:               32K
Cache L1i:               32K
Cache L2:                256K
Cache L3:                46080K
Procesory węzła NUMA 0:  0
Flagi:                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault invpcid_single pti fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt
```

**Comment by jasonbosco** [comment]
Could you check if the node ran out of disk space at some point? If so, then the only way to recover it is to add more disk space and restart the process.

**Comment by jasiek-net** [comment]
I checked disk usage, it looks okey
```
devtmpfs        1,1G     0  1,1G   0% /dev
tmpfs           1,1G     0  1,1G   0% /dev/shm
tmpfs           1,1G  541k  1,1G   1% /run
tmpfs           1,1G     0  1,1G   0% /sys/fs/cgroup
/dev/xvda1      8,6G  4,5G  4,2G  52% /
tmpfs           207M     0  207M   0% /run/user/1000
overlay         8,6G  4,5G  4,2G  52% /var/lib/docker/overlay2/3573248e9fa4b9e2767576b9d803498d9b8580a168f2e692e15011c1459a527b/merged
```

Every time I scrap new data, I delete previous data from typesense, so it shouldn't keep more space.

**Comment by jasonbosco** [comment]
Could you share the logs from the time you start a fresh instance of Typesense, to when it goes into this state? 

If the log file is too big to share as a zip here, feel free to email it to contact at typesense dot org

**Comment by andreypetruk1** [comment]
Hi. I had the same problem, but in my case it went to error state after 6 week. Version typesense/typesense:0.23.1, single instance. Did not see any issues with disk space or RAM. After restart it was fixed.
There are logs of the moment when it goes to error state

> 15:42:56.601305   133 batched_indexer.cpp:250] Running GC for aborted requests, req map size: 0
15:42:59.244760   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
15:42:59.244837   173 raft_server.h:60] Peer refresh succeeded!
15:43:09.245932   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
15:43:09.246045   175 raft_server.h:60] Peer refresh succeeded!
15:43:19.247015   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
15:43:19.247124   171 raft_server.h:60] Peer refresh succeeded!
15:43:29.248172   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
15:43:29.248296   171 raft_server.h:60] Peer refresh succeeded!
15:43:39.249282   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
15:43:39.249399   171 raft_server.h:60] Peer refresh succeeded!
16:12:19.448117   132 node.cpp:866] node default_group:8107:8108 is in state ERROR, can't reset_peer
16:12:25.780254   133 batched_indexer.cpp:250] Running GC for aborted requests, req map size: 0
16:12:29.449424   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
16:12:29.449450   132 raft_server.cpp:558] Single-node with no leader. Resetting peers.
16:12:29.449455   132 node.cpp:866] node default_group:8107:8108 is in state ERROR, can't reset_peer
16:12:39.450417   132 raft_server.cpp:534] Term: 13, last_index index: 2701, committed_index: 2701, known_applied_index: 2701, applying_index: 0, queued_writes: 0, pending_queue_size: 0, local_sequence: 36073
16:12:39.450448   132 raft_server.cpp:558] Single-node with no leader. Resetting peers.

**Comment by MeikelLP** [comment]
I have this problem locally running in a docker container for my dev environment. I'm currently evaluating Typesense and this is a total showstopper if Typesense suddenly stops working in production.

```
2023-04-19 19:35:24 I20230419 17:35:24.041591   601 raft_server.h:60] Peer refresh succeeded!
2023-04-19 19:35:34 I20230419 17:35:34.042662   454 raft_server.cpp:546] Term: 5, last_index index: 5997, committed_index: 5997, known_applied_index: 5997, applying_index: 0, queued_writes: 1, pending_queue_size: 0, local_sequence: 29842
2023-04-19 19:35:34 I20230419 17:35:34.042716   454 raft_server.cpp:863] Snapshot timer is active, current_ts: 1681925734, last_snapshot_ts: 1681913559
2023-04-19 19:35:34 I20230419 17:35:34.042724   454 node.cpp:911] node default_group:172.24.0.2:8107:8108 starts to do snapshot
2023-04-19 19:35:34 I20230419 17:35:34.042760   600 raft_server.h:60] Peer refresh succeeded!
2023-04-19 19:35:34 I20230419 17:35:34.051291   590 raft_server.cpp:446] on_snapshot_save
2023-04-19 19:35:34 I20230419 17:35:34.051323   590 batched_indexer.cpp:396] Serialized 0 in-flight requests for snapshot.
2023-04-19 19:35:34 W20230419 17:35:34.088055   590 store.h:349] Checkpoint CreateCheckpoint failed at snapshot path: /data/state/snapshot/temp/db_snapshot, msg:Corruption: block checksum mismatch: stored = 2324967102, computed = 2998824592  in /data/db/000153.sst offset 207671 size 3727
2023-04-19 19:35:34 E20230419 17:35:34.088088   590 raft_server.cpp:468] Failure during checkpoint creation, msg:Corruption: block checksum mismatch: stored = 2324967102, computed = 2998824592  in /data/db/000153.sst offset 207671 size 3727
2023-04-19 19:35:34 I20230419 17:35:34.088119   590 raft_server.cpp:390] save_snapshot called
2023-04-19 19:35:34 I20230419 17:35:34.089308   590 snapshot.cpp:516] Deleting /data/state/snapshot/temp
2023-04-19 19:35:34 W20230419 17:35:34.093549   590 snapshot_executor.cpp:220] node default_group:172.24.0.2:8107:8108 fail to close writer
2023-04-19 19:35:34 E20230419 17:35:34.093613   590 raft_server.cpp:925] Timed snapshot failed, error: Checkpoint creation failure., code: 5
2023-04-19 19:35:34 E20230419 17:35:34.093616   601 raft_server.h:269] Met peering error {type=SnapshotError, error_code=5, error_text=`Fail to save snapshot'}
2023-04-19 19:35:34 W20230419 17:35:34.093626   601 node.cpp:1264] node default_group:172.24.0.2:8107:8108 got error={type=SnapshotError, error_code=5, error_text=`Fail to save snapshot'}
2023-04-19 19:35:34 I20230419 17:35:34.093670   601 replicator.cpp:1475] Group default_group Fail to find the next candidate
2023-04-19 19:35:34 I20230419 17:35:34.093695   601 raft_server.h:261] Node stepped down : Raft node(leader or candidate) is in error.
2023-04-19 19:35:44 I20230419 17:35:44.052564   454 raft_server.cpp:546] Term: 5, last_index index: 5998, committed_index: 5997, known_applied_index: 5997, applying_index: 0, queued_writes: 1, pending_queue_size: 0, local_sequence: 29842
2023-04-19 19:35:44 W20230419 17:35:44.052597   454 raft_server.cpp:570] Single-node with no leader. Resetting peers.
2023-04-19 19:35:44 W20230419 17:35:44.052603   454 node.cpp:866] node default_group:172.24.0.2:8107:8108 is in state ERROR, can't reset_peer
2023-04-19 19:35:54 I20230419 17:35:54.053864   454 raft_server.cpp:546] Term: 5, last_index index: 5998, committed_index: 5997, known_applied_index: 5997, applying_index: 0, queued_writes: 1, pending_queue_size: 0, local_sequence: 29842
```

From what I can see in the logs it sounds like the snapshop (saving to disk) fails due to corruption (?) and thus causes the (single) node to no longer respond

**Comment by kishorenc** [comment]
As with all databases, corruption can happen to due to either hardware or software error. A few observations:

- The error happens because the snapshot process finds one of the rocksdb file checksums to mismatch. 
- So far every instance of this issue seems to happen when Typesense is run on Docker. We run thousands of nodes on Typesense Cloud and we've never faced any corruption issue (we don't use Docker).
- We have to ascertain whether the issue is with Docker or is a hardware/software fault that seems to happen more often with Docker

For now, to recover from this issue, just delete the `db` directory inside the Typesense data-dir so Typesense can restore the last good snapshot and recover. You don't need to reindex the data.

**Comment by MeikelLP** [comment]
Yes deleting the `db` healed the application tho actual data was lost. 
Sadly I had 5000 entries. After that corruption I only had about 40.

If you say this is Docker's fault and not Typsense's then I will have to create a re-index feature in my app because Typesense is clearly not stable enough. Fine for me because it's just a search layer on top of the stable database. This is the first time I ever heard someone put such an issue to Docker. Never had such an issue with any Database or even Elastic search in Docker.

**Comment by kishorenc** [comment]
> This is the first time I ever heard someone put such an issue to Docker. 

Please read what I wrote carefully. I said that it could be an issue related to Docker given my experience of running clusters without it. If there is such a stability issue we would have come across that over thousands of native deployments (we serve 750M+ requests on Typesense Cloud every month). 

For us to identify what's going on, our ability to reproduce it is crucial: if you can kindly see if it is easy to reproduce, please let me know so I can take a closer look.

**Comment by kishorenc** [comment]
@MeikelLP Did you try restarting the Docker container to see if the service recovered before deleting the `db` dir?

**Comment by MeikelLP** [comment]
Yes it did recover but soon ran into the problem again (after about a minute) probably once it tried to save another snapshot.

**Comment by swood** [comment]
I have similar problem:

> W20240403 23:15:45.719142   290 store.h:350] Checkpoint CreateCheckpoint failed at snapshot path: /data/state/snapshot/temp/db_snapshot, msg:IO error: While fdatasync: /data/db/859198.sst: Input/output error
E20240403 23:15:45.719158   290 raft_server.cpp:478] Failure during checkpoint creation, msg:IO error: While fdatasync: /data/db/859198.sst: Input/output error
I20240403 23:15:45.719184   290 raft_server.cpp:398] save_snapshot called
I20240403 23:15:45.719226   290 snapshot.cpp:519] Deleting /data/state/snapshot/temp
W20240403 23:15:45.719281   290 snapshot_executor.cpp:220] node default_group:172.16.2.5:8107:8108 fail to close writer
E20240403 23:15:45.720293   290 raft_server.cpp:994] Timed snapshot failed, error: Checkpoint creation failure., code: 5
E20240403 23:15:45.720752   291 raft_server.h:281] Met peering error {type=SnapshotError, error_code=5, error_text=`Fail to save snapshot'}
W20240403 23:15:45.720794   291 node.cpp:1311] node default_group:172.16.2.5:8107:8108 got error={type=SnapshotError, error_code=5, error_text=`Fail to save snapshot'}
I20240403 23:15:45.720834   291 replicator.cpp:1499] Group default_group Fail to find the next candidate
I20240403 23:15:45.721267   291 raft_server.h:273] Node stepped down : Raft node(leader or candidate) is in error.
I20240403 23:15:46.259994   263 batched_indexer.cpp:288] Running GC for aborted requests, req map size: 0
E20240403 23:15:49.720614   299 http_client.cpp:165] CURL timeout. Time taken: 4.00118, method: POST, url: http://172.16.2.5:8108/health
I20240403 23:15:49.720713   299 raft_server.cpp:770] Dummy write to http://172.16.2.5:8108/health, status = 408, response =
I20240403 23:15:49.720731   299 raft_server.cpp:449] save_snapshot done

For some reason, the snapshot process was unsuccessful. After that, the Typesense instance was unreachable until I restarted the container.

I am curious: why is the process can't skip a snapshot and try to make it later?
Typesense version: 0.25.2.rc10

**Comment by itinance** [comment]
I face the same issue and after "surviving" this period, it couldn't perform any search any more: https://github.com/typesense/typesense/issues/1993
